src/openmc_heat_driver.cpp

```diff
@@ -214,42 +368,73 @@ void OpenmcHeatDriver::set_heat_source()
   }
 }
 
+void OpenmcHeatDriver::update_density()
+{
+  if (this->has_global_coupling_data()) {
+    std::copy(densities_.begin(), densities_.end(), densities_prev_.begin());
+  }
+
+  densities_ = heat_driver_->density();
+  int solid_offset = heat_driver_->n_pins_ * heat_driver_->n_rings() * heat_driver_->n_axial_;
+
+  for (int i = 0; i < n_fluid_cells_; ++i) {
+    int index = i + n_solid_cells_;
+    const auto& c = openmc_driver_->cells_[index];
+    const auto& elems = cell_inst_to_elem_[index];
+
+    double average_rho = 0.0;
+    double total_vol = 0.0;
+    for (int elem_index : elems) {
+      double vol = heat_driver_->z_(elem_index + 1) - heat_driver_->z_(elem_index);
+      average_rho += densities_(elem_index + solid_offset);
```

Comment:
Should be `* vol`, no?

Reply:
Good catch! I guess I was only testing it with one-to-one mappings since I haven't thought of a good way to test whether all OpenMC fluid cells are included